lib/phil/charlie_cards/importer.ex

```diff
@@ -0,0 +1,61 @@
+defmodule Phil.CharlieCards.Importer do
+  @moduledoc """
+  Imports CharlieCard inventory into Phil.
+  """
+
+  import Ecto.Query, only: [from: 2]
+
+  alias NimbleCSV.RFC4180, as: CSV
+
+  alias Phil.CharlieCards.CharlieCard
+  alias Phil.Products.Product
+  alias Phil.Repo
+
+  @media %{"2" => :ticket, "5" => :card}
+
+  @doc """
+  Given the path of a file, this function attempts to import CharlieCards.
+
+  It returns a map of successes and failures like %{error: [], ok: []}.
+  Error entries are changesets and success entries are the inserted struct.
+  """
+  def import(path) do
+    path
+    |> File.stream!()
+    |> CSV.parse_stream()
+    |> Stream.map(&row_to_changeset/1)
+    |> Stream.chunk_every(100)
+    |> Task.async_stream(&insert_batch/1, max_concurrency: 10, ordered: false)
+    |> Enum.reduce([], fn {:ok, results}, acc -> acc ++ results end)
+    |> Enum.group_by(fn {status, _} -> status end, fn {_, value} -> value end)
+  end
+
+  defp insert_batch(changesets) do
```

Comment:
it'd probably be better to do an insert_all on a bunch of them, is that not possible?

Reply:
`insert_all` doesn't work with Changesets, does it? The idea is to give the user back some actionable information on a per-row basis.

Reply:
Well you can still do the changeset validations per row, but then if that passes transform everything into a map suitable for insert_all. Eg something like:

```elixir
case validate_changest(ch) do
  {:error, changeset} -> 
  {:ok, changeset} -> Ecto.Changeset.apply_action(:insert) |> Map.take([...]) |> Map.merge(timetamps)
end
```

Reply:
Will that work with validating unique constraints and foreign key constraints?

Reply:
I'll also note that this takes 2 seconds to import 3k cards. And, it's going to be used every few months. So, I'm not sure it's worth re-thinking.

Reply:
So there is an unsafe validation but obvs unique constraints really only the DB can guarantee. But this is why I am asking about upserts

Reply:
though I'm not sure in your example that anything got inserted as the id is nil in what you pasted

Reply:
Right, but what I'm saying is that if you have two rows and you duplicated the serial number, it will insert the first then upsert it. You'll end up with one card and you won't know you made a mistake.

Reply:
yes, which is why I'm asking if that's an expected failiure case? Do we have evidence that that is possible, or is the only possibility of duplicated rows if someone re-uploads the same CSV

Reply:
Super possible. I was given duplicate serial numbers and duplicate ticket type ids just in the test data.

Reply:
nice okay cool thanks. And the rest of the data for duplicated rows was different? Or was it the exact same row there twice?

Reply:
All different. Like, for the products there were like 8 different products using the same ticket type id as stored value and they all had stored value in the name. For inventory, the original file had a ton of duplicate serial numbers that were also just wrong. There is a very manual component to the process.

Reply:
wow! good to know thanks